pills/17-nixpkgs-overriding-packages.xml

```diff
@@ -51,23 +51,20 @@
     </para>
     <screen><xi:include href="./17/fix-function.txt" parse="text" /></screen>
     <para>
-      It's a function that accepts a function <literal>f</literal>, calls <literal>f result</literal> on the result just returned by <literal>f result</literal> and returns it. In other words it's <literal>f(f(f(....</literal>
+      It's a function that accepts a function <literal>f</literal> and returns a function that recurses over f's arguments. Why is this useful? Because it plays nicely with Nix' laziness: it turns any function that otherwise would happily return (at least some) of its arguments unevaluated, into a function that recursively evaluates its arguments until there is nothing left to evaluate. To fix ideas (pun intended!) consider:
     </para>
+    <screen><xi:include href="./17/fix-function-sum.txt" parse="text" /></screen>
     <para>
-      At first sight, it's an infinite loop. With lazy evaluation it isn't, because the call is done only when needed.
+        What is the main difference between <literal>fix (const (sum))</literal> and <literal>sum</literal>? Thanks to <literal>fix</literal>, the function recursively evaluates its arguments until there is no unevaluated value left (unevaluated values are called <literal>thunks</literal>). If at any point in the recursion an argument is found to be a thunk, it is evaluated and the result is passed to the next function application in the recursion.
```

Comment:
Doesn't `fix (const sum)` simply evaluate to `sum`?

I've been struggling with trying to understand `fix` and I already spent a couple hours on this, but can't see where the error is if this is not correct:

```
fix f = f( f( f( ... )))

fix (const sum) = ( const sum) ( (const sum) ( ... ))
```

where `const sum` can be though of as a wrapper:

```
fix (x: sum) = (x: sum) ( (x: sum) ( ... )) = sum
    ---f---- = ---f---- ( ---f---- ( ... ))
```

To belabor the point, input `x` is literally ignored:

```
( fix ( x: sum )) 1 2

   ┏━━━━━━━━━━━━━━━━━━┓
   ∨          --------┃-------
( (x: sum) (  (x: sum) ( ... )  )  ) 1 2
  -----------┃-------------------
             ┃
             ∨
         (  sum  ) 1 2
```